### ISSUE TITLE
TFLint fails if a provider is overridden

### ISSUE THREAD

**Issue by gustavderdrache** [terraform-linters/tflint#1002]
I have a Terraform project where the AWS provider uses `assume_role` (see the [provider docs](https://registry.terraform.io/providers/hashicorp/aws/latest/docs#assume_role-configuration-block)). I can't use the role locally, so I use an override file to let me run `terraform validate` and `terraform plan` for sanity checking before I push my branch up.

However, it appears that TFLint has issues with this override. Terraform is able to handle it just fine, so it seems like the issue lies in how TFLint handles the override file.

Here is the (simplified) code for both files:

```hcl
# providers.tf
provider "aws" {
  region = "us-east-1"

  assume_role {
    role_arn = "arn:aws:iam::123412341234:role/ExampleRole"
  }
}

# providers_override.tf
provider "aws" {
  region = "us-east-1"

  assume_role {
    role_arn = null
  }
}
```

The error message is pretty cryptic, unfortunately. Here is the full output with TFLint's debug log:

```
$ docker run --rm -v $(pwd):/data:ro -t wata727/tflint:0.22.0 --loglevel=debug
18:29:27 config.go:98: [INFO] Load config: .tflint.hcl
18:29:27 config.go:110: [INFO] Default config file is not found. Ignored
18:29:27 config.go:119: [INFO] Load fallback config: /root/.tflint.hcl
18:29:27 config.go:127: [INFO] Fallback config file is not found. Ignored
18:29:27 config.go:129: [INFO] Use default config
18:29:27 option.go:54: [DEBUG] CLI Options
18:29:27 option.go:55: [DEBUG]   Module: false
18:29:27 option.go:56: [DEBUG]   DeepCheck: false
18:29:27 option.go:57: [DEBUG]   Force: false
18:29:27 option.go:58: [DEBUG]   IgnoreModules: map[string]bool{}
18:29:27 option.go:59: [DEBUG]   EnableRules: []string(nil)
18:29:27 option.go:60: [DEBUG]   DisableRules: []string(nil)
18:29:27 option.go:61: [DEBUG]   Only: []string(nil)
18:29:27 option.go:62: [DEBUG]   Varfiles: []string{}
18:29:27 option.go:63: [DEBUG]   Variables: []string{}
18:29:27 loader.go:57: [INFO] Initialize new loader
18:29:27 loader.go:82: [INFO] Load configurations under .
18:29:27 loader.go:90: [INFO] Module inspection is disabled. Building a root module without children...
18:29:27 loader.go:170: [INFO] Load values files
18:29:27 runner.go:52: [INFO] Initialize new runner for root
18:29:27 provider.go:87: [INFO] Prepare rules
18:29:27 provider.go:123: [INFO]   768 rules enabled
18:29:27 runner_walk.go:36: [DEBUG] Walk `aws_s3_bucket.alb_logs.acl` attribute
18:29:27 runner_eval.go:67: [WARN] Unevaluable expression found in s3.tf:2; TFLint ignores an unevaluable expression.
18:29:27 runner_walk.go:20: [WARN] Skip walking `aws_s3_bucket.index_s3` because it may not be created
18:29:27 runner_eval.go:67: [WARN] Unevaluable expression found in s3.tf:2; TFLint ignores an unevaluable expression.
18:29:27 runner_walk.go:20: [WARN] Skip walking `aws_s3_bucket.index_s3` because it may not be created
Failed to check `terraform_deprecated_interpolation` rule. 2 error(s) occurred:

Error: Unexpected "assume_role" block

  on providers.tf line 31, in provider "aws":
  31:   assume_role {

Blocks are not allowed here.

Error: Unexpected "assume_role" block

  on providers_override.tf line 4, in provider "aws":
   4:   assume_role {

Blocks are not allowed here.
```

## Version

* TFLint: v0.22.0 (via Docker)
* Terraform: v0.13.5

**Comment by bendrucker** [comment]
Interesting, TFLint calls Terraform as a library and I thought it handled overrides there. But I've never used them nor looked carefully. 

FYI Terraform should not be configuring the provider during `validate` so it should not matter whether you could actually assume the role. It does configure the provider for a `plan`.

**Comment by gustavderdrache** [comment]
I did some extra experimentation to see if I could narrow down the behavior. It's a bit odd, but it seems to only affect blocks in any declaration - not just providers.

I tested both combinations of adding a block to an S3 bucket:

1. Versioning in the resource file:

   ```hcl
   # s3.tf
   resource "aws_s3_bucket" "bucket" {
     bucket = "bucket"

     versioning {
       enabled = true
     }
   }
   ```

   ```hcl
   # s3_override.tf
   resource "aws_s3_bucket" "bucket" {}
   ```

   Error output:

   ```
   $ docker run --rm -v $(pwd):/data:ro -t wata727/tflint:0.22.0 --loglevel=debug
   15:55:21 config.go:98: [INFO] Load config: .tflint.hcl
   15:55:21 config.go:110: [INFO] Default config file is not found. Ignored
   15:55:21 config.go:119: [INFO] Load fallback config: /root/.tflint.hcl
   15:55:21 config.go:127: [INFO] Fallback config file is not found. Ignored
   15:55:21 config.go:129: [INFO] Use default config
   15:55:21 option.go:54: [DEBUG] CLI Options
   15:55:21 option.go:55: [DEBUG]   Module: false
   15:55:21 option.go:56: [DEBUG]   DeepCheck: false
   15:55:21 option.go:57: [DEBUG]   Force: false
   15:55:21 option.go:58: [DEBUG]   IgnoreModules: map[string]bool{}
   15:55:21 option.go:59: [DEBUG]   EnableRules: []string(nil)
   15:55:21 option.go:60: [DEBUG]   DisableRules: []string(nil)
   15:55:21 option.go:61: [DEBUG]   Only: []string(nil)
   15:55:21 option.go:62: [DEBUG]   Varfiles: []string{}
   15:55:21 option.go:63: [DEBUG]   Variables: []string{}
   15:55:21 loader.go:57: [INFO] Initialize new loader
   15:55:21 loader.go:82: [INFO] Load configurations under .
   15:55:21 loader.go:90: [INFO] Module inspection is disabled. Building a root module without children...
   15:55:21 loader.go:170: [INFO] Load values files
   15:55:21 runner.go:52: [INFO] Initialize new runner for root
   15:55:21 provider.go:87: [INFO] Prepare rules
   15:55:21 provider.go:123: [INFO]   768 rules enabled
   Failed to check `terraform_deprecated_interpolation` rule. 1 error(s) occurred:

   Error: Unexpected "versioning" block

     on s3.tf line 4, in resource "aws_s3_bucket" "bucket":
      4:   versioning {

   Blocks are not allowed here.
   ```

2. Versioning in the override:

   ```hcl
   # s3.tf
   resource "aws_s3_bucket" "bucket" {
     bucket = "bucket"
   }
   ```

   ```hcl
   # s3_override.tf
   resource "aws_s3_bucket" "bucket" {
     versioning {
       enabled = true
     }
   }
   ```

   Error output:

   ```
   $ docker run --rm -v $(pwd):/data:ro -t wata727/tflint:0.22.0 --loglevel=debug
   15:59:42 config.go:98: [INFO] Load config: .tflint.hcl
   15:59:42 config.go:110: [INFO] Default config file is not found. Ignored
   15:59:42 config.go:119: [INFO] Load fallback config: /root/.tflint.hcl
   15:59:42 config.go:127: [INFO] Fallback config file is not found. Ignored
   15:59:42 config.go:129: [INFO] Use default config
   15:59:42 option.go:54: [DEBUG] CLI Options
   15:59:42 option.go:55: [DEBUG]   Module: false
   15:59:42 option.go:56: [DEBUG]   DeepCheck: false
   15:59:42 option.go:57: [DEBUG]   Force: false
   15:59:42 option.go:58: [DEBUG]   IgnoreModules: map[string]bool{}
   15:59:42 option.go:59: [DEBUG]   EnableRules: []string(nil)
   15:59:42 option.go:60: [DEBUG]   DisableRules: []string(nil)
   15:59:42 option.go:61: [DEBUG]   Only: []string(nil)
   15:59:42 option.go:62: [DEBUG]   Varfiles: []string{}
   15:59:42 option.go:63: [DEBUG]   Variables: []string{}
   15:59:42 loader.go:57: [INFO] Initialize new loader
   15:59:42 loader.go:82: [INFO] Load configurations under .
   15:59:42 loader.go:90: [INFO] Module inspection is disabled. Building a root module without children...
   15:59:42 loader.go:170: [INFO] Load values files
   15:59:42 runner.go:52: [INFO] Initialize new runner for root
   15:59:42 provider.go:87: [INFO] Prepare rules
   15:59:42 provider.go:123: [INFO]   768 rules enabled
   Failed to check `terraform_deprecated_interpolation` rule. 1 error(s) occurred:

   Error: Unexpected "versioning" block

     on s3_override.tf line 2, in resource "aws_s3_bucket" "bucket":
      2:   versioning {

   Blocks are not allowed here.
   ```

So I did a little more testing, and found that TFLint _is_ at least recognizing the overrides as overrides.

```hcl
# s3.tf
resource "aws_s3_bucket" "bucket" {
  bucket = "bucket"
}
```

```hcl
# s3_override.tf
resource "aws_s3_bucket" "bucketx" {}
```

TFLint/Terraform complains appropriately about `bucketx` not being a valid resource:

```
$ docker run --rm -v $(pwd):/data:ro -t wata727/tflint:0.22.0 --loglevel=debug
15:58:01 config.go:98: [INFO] Load config: .tflint.hcl
15:58:01 config.go:110: [INFO] Default config file is not found. Ignored
15:58:01 config.go:119: [INFO] Load fallback config: /root/.tflint.hcl
15:58:01 config.go:127: [INFO] Fallback config file is not found. Ignored
15:58:01 config.go:129: [INFO] Use default config
15:58:01 option.go:54: [DEBUG] CLI Options
15:58:01 option.go:55: [DEBUG]   Module: false
15:58:01 option.go:56: [DEBUG]   DeepCheck: false
15:58:01 option.go:57: [DEBUG]   Force: false
15:58:01 option.go:58: [DEBUG]   IgnoreModules: map[string]bool{}
15:58:01 option.go:59: [DEBUG]   EnableRules: []string(nil)
15:58:01 option.go:60: [DEBUG]   DisableRules: []string(nil)
15:58:01 option.go:61: [DEBUG]   Only: []string(nil)
15:58:01 option.go:62: [DEBUG]   Varfiles: []string{}
15:58:01 option.go:63: [DEBUG]   Variables: []string{}
15:58:01 loader.go:57: [INFO] Initialize new loader
15:58:01 loader.go:82: [INFO] Load configurations under .
15:58:01 loader.go:85: [ERROR] s3_override.tf:1,1-35: Missing resource to override; There is no aws_s3_bucket resource named "bucketx". An override file can only override a resource block defined in a primary configuration file.
Failed to load configurations. 1 error(s) occurred:

Error: Missing resource to override

  on s3_override.tf line 1, in resource "aws_s3_bucket" "bucketx":
   1: resource "aws_s3_bucket" "bucketx" {}

There is no aws_s3_bucket resource named "bucketx". An override file can only override a resource block defined in a primary configuration file.
```

**Comment by bendrucker** [comment]
Good info, definitely makes sense that it's not limited to providers. What's odd here is that we're relying on Terraform libraries to load a directory of configuration files and turn it into a module object. 

https://pkg.go.dev/github.com/hashicorp/terraform/configs#Parser.LoadConfigDir

That method definitely handles overrides. You're seeing that behavior in your last output, where `loader.go` tries to load/parse a configuration with invalid overrides.

So something about the way overrides are processed seems to mean the parsed HCL is different, resulting in evaluation errors at runtime. 

It'll definitely be easiest to investigate this by debugging, if you're so inclined. Otherwise I can try to look at it soon.

**Comment by gustavderdrache** [comment]
I haven't debugged Go before but I'm happy to take a crack at it this weekend to see if I can spot anything. If I don't figure it out, I'll let you know so you can take over.
